crates/sui-rest-api/src/transactions/resolve.rs

```diff
@@ -0,0 +1,599 @@
+        )),
+    }
+    .pipe(Ok)
+}
+
+/// Given an particular input argument, find all of its uses.
+///
+/// The returned iterator contains all commands where the arguement is used an an optional index
+/// for where the argument is used in that command.
+fn find_arg_uses(
+    arg_idx: usize,
+    commands: &[Command],
+) -> impl Iterator<Item = (&Command, Option<usize>)> {
+    commands.iter().filter_map(move |command| {
+        match command {
+            Command::MoveCall(move_call) => move_call
+                .arguments
+                .iter()
+                .position(|elem| matches_input_arg(*elem, arg_idx))
+                .map(Some),
+            Command::TransferObjects(transfer_objects) => transfer_objects
+                .objects
+                .iter()
+                .position(|elem| matches_input_arg(*elem, arg_idx))
+                .map(Some),
+            Command::SplitCoins(split_coins) => {
+                matches_input_arg(split_coins.coin, arg_idx).then_some(None)
+            }
+            Command::MergeCoins(merge_coins) => {
+                if matches_input_arg(merge_coins.coin, arg_idx) {
+                    Some(None)
+                } else {
+                    merge_coins
+                        .coins_to_merge
+                        .iter()
+                        .position(|elem| matches_input_arg(*elem, arg_idx))
+                        .map(Some)
+                }
+            }
+            Command::Publish(_) => None,
+            Command::MakeMoveVector(make_move_vector) => make_move_vector
+                .elements
+                .iter()
+                .position(|elem| matches_input_arg(*elem, arg_idx))
+                .map(Some),
+            Command::Upgrade(upgrade) => matches_input_arg(upgrade.ticket, arg_idx).then_some(None),
+        }
+        .map(|x| (command, x))
+    })
+}
+
+fn matches_input_arg(arg: Argument, arg_idx: usize) -> bool {
+    matches!(arg, Argument::Input(idx) if idx as usize == arg_idx)
+}
+
+/// Estimate the gas budget using the gas_cost_summary from a previous DryRun
+///
+/// The estimated gas budget is computed as following:
+/// * the maximum between A and B, where:
+///     A = computation cost + GAS_SAFE_OVERHEAD * reference gas price
+///     B = computation cost + storage cost - storage rebate + GAS_SAFE_OVERHEAD * reference gas price
+///     overhead
+///
+/// This gas estimate is computed similarly as in the TypeScript SDK
+fn estimate_gas_budget_from_gas_cost(
+    gas_cost_summary: &GasCostSummary,
+    reference_gas_price: u64,
+) -> u64 {
+    const GAS_SAFE_OVERHEAD: u64 = 1000;
+
+    let safe_overhead = GAS_SAFE_OVERHEAD * reference_gas_price;
+    let computation_cost_with_overhead = gas_cost_summary.computation_cost + safe_overhead;
+
+    let gas_usage = gas_cost_summary.net_gas_usage() + safe_overhead as i64;
+    computation_cost_with_overhead.max(if gas_usage < 0 { 0 } else { gas_usage as u64 })
+}
+
+fn select_gas(
```

Comment:
This just does coin selection but does not smash gas, right?

Reply:
any gas coins selected are smashed during execution

Reply:
Give me your email address i have some questions for you